src/databricks/labs/ucx/source_code/redash.py

```diff
@@ -1,24 +1,172 @@
+import logging
+from collections.abc import Iterator
+from dataclasses import replace
+
+from databricks.labs.blueprint.tui import Prompts
+
 from databricks.sdk import WorkspaceClient
-from databricks.sdk.service.sql import Query
+from databricks.sdk.service.sql import Query, Dashboard
+from databricks.sdk.errors.platform import DatabricksError, NotFound
+
+from databricks.labs.ucx.hive_metastore.migration_status import MigrationIndex
+from databricks.labs.ucx.source_code.base import CurrentSessionState
+from databricks.labs.ucx.source_code.queries import FromTable
 
-from databricks.labs.ucx.source_code.base import Fixer
+logger = logging.getLogger(__name__)
 
 
 class Redash:
-    def __init__(self, fixer: Fixer, ws: WorkspaceClient):
-        self._fixer = fixer
+    MIGRATED_TAG = "migrated by UCX"
+    BACKUP_TAG = "backup by UCX"
+
+    def __init__(self, index: MigrationIndex, ws: WorkspaceClient, backup_path: str):
+        self._index = index
         self._ws = ws
+        self._backup_path = backup_path + "/backup_queries"
+
+    def migrate_dashboards(self, dashboard_id: str | None = None):
+        for dashboard in self._list_dashboards(dashboard_id):
+            assert dashboard.id is not None
+            if dashboard.tags is not None and self.MIGRATED_TAG in dashboard.tags:
+                logger.debug(f"Dashboard {dashboard.name} already migrated by UCX")
+                continue
+            for query in self.get_queries_from_dashboard(dashboard):
+                self._fix_query(query)
+            self._ws.dashboards.update(dashboard.id, tags=self._get_migrated_tags(dashboard.tags))
 
-    def fix(self, query: Query):
+    def revert_dashboards(self, dashboard_id: str | None = None):
+        for dashboard in self._list_dashboards(dashboard_id):
+            assert dashboard.id is not None
+            if dashboard.tags is None or self.MIGRATED_TAG not in dashboard.tags:
+                logger.debug(f"Dashboard {dashboard.name} was not migrated by UCX")
+                continue
+            for query in self.get_queries_from_dashboard(dashboard):
+                self._revert_query(query)
+            self._ws.dashboards.update(dashboard.id, tags=self._get_original_tags(dashboard.tags))
+
+    def _list_dashboards(self, dashboard_id: str | None) -> list[Dashboard]:
+        try:
+            if dashboard_id is None:
+                return list(self._ws.dashboards.list())
+            return [self._ws.dashboards.get(dashboard_id)]
+        except DatabricksError as e:
+            logger.error(f"Cannot list dashboards: {e}")
```

Comment:
```suggestion
            logger.warning(f"Cannot list dashboards: {e}")
```

Reply:
we need to fail the job if dashboards cannot be listed